streams/test-utils/src/main/java/org/apache/kafka/streams/TestOutputTopic.java

```diff
@@ -0,0 +1,200 @@
+ */
+public class TestOutputTopic<K, V> {
+    private final TopologyTestDriver driver;
+    private final String topic;
+    private final Deserializer<K> keyDeserializer;
+    private final Deserializer<V> valueDeserializer;
+
+    /**
+     * Create a test output topic to read messages from
+     *
+     * @param driver            TopologyTestDriver to use
+     * @param topicName         the topic name used
+     * @param keyDeserializer   the key deserializer
+     * @param valueDeserializer the value deserializer
+     */
+    TestOutputTopic(final TopologyTestDriver driver,
+                    final String topicName,
+                    final Deserializer<K> keyDeserializer,
+                    final Deserializer<V> valueDeserializer) {
+        Objects.requireNonNull(driver, "TopologyTestDriver cannot be null");
+        Objects.requireNonNull(topicName, "topicName cannot be null");
+        Objects.requireNonNull(keyDeserializer, "keyDeserializer cannot be null");
+        Objects.requireNonNull(valueDeserializer, "valueDeserializer cannot be null");
+        this.driver = driver;
+        this.topic = topicName;
+        this.keyDeserializer = keyDeserializer;
+        this.valueDeserializer = valueDeserializer;
+    }
+
+
+    /**
+     * Read one Record from output topic and return value only.
+     * <p>
+     * Note. The key and header is not available
+     *
+     * @return Next value for output topic
+     */
+    public V readValue() {
+        final TestRecord<K, V> record = readRecord();
+        return record.value();
+    }
+
+    /**
+     * Read one Record KeyValue from output topic.
+     *
+     * @return Next output as KeyValue
+     */
+    public KeyValue<K, V> readKeyValue() {
+        final TestRecord<K, V> record = readRecord();
+        return new KeyValue<>(record.key(), record.value());
+    }
+
+    /**
+     * Read one Record from output topic.
+     *
+     * @return Next output as TestRecord
+     */
+    public TestRecord<K, V> readRecord() {
+        return driver.readRecord(topic, keyDeserializer, valueDeserializer);
+    }
+
+    /**
+     * Read output to List.
+     * If the existing key is modified, it can appear twice in output, but replaced in map
+     *
+     * @return Map of output by key
+     */
+    public List<TestRecord<K, V>> readRecordsToList() {
+        final List<TestRecord<K, V>> output = new LinkedList<>();
+        while (!isEmpty()) {
+            output.add(readRecord());
+        }
+        return output;
+    }
+
+
+    /**
+     * Read output to map.
+     * If the existing key is modified, it can appear twice in output, but replaced in map
+     *
+     * @return Map of output by key
+     */
+    public Map<K, V> readKeyValuesToMap() {
+        final Map<K, V> output = new HashMap<>();
+        TestRecord<K, V> outputRow;
+        while (!isEmpty()) {
+            outputRow = readRecord();
+            output.put(outputRow.key(), outputRow.value());
```

Comment:
We should point out in the JavaDocs, that null-values don't have delete semantics! Ie, if the last update to a key is a delete/tombstone, the key will still be in the map (with null-value).

Also, I think we should not allow `null` keys, but throw an exception for this case.

Reply:
Added